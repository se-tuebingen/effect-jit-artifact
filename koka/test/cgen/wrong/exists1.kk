// tests codegeneration for existentials
type expr<a> {
  Lit( x : a )
  If( pred : expr<bool>, t : expr<a>, e : expr<a> )
  App<b>( f : expr<b -> a>, arg : expr<b> )
}

fun is-positive( i : int ) : bool {
  i.is-pos
}

val bexpr : expr<bool> =  App(Lit(is-positive),Lit(42))

fun eval(expr : expr<a>) : a {
  match(expr) {
    Lit(x)    -> x
    If(p,t,e) -> if (p.eval) then t.eval else e.eval
    App(f,x)  -> (f.eval)(True)
  }
}

fun main() {
  println(eval(bexpr))
}
