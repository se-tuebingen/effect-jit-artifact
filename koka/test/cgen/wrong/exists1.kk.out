test/cgen/wrong/exists1@kk(18,27): type error: True does not match the argument types
 context : True
 term : True
 inferred type: bool
 expected type: $a